File: app/src/main/java/com/notai/ui/viewmodel/TagViewModel.java
package com.notai.ui.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.notai.NoteAIApplication;
import com.notai.data.model.Tag;
import com.notai.data.repository.TagRepository;

import java.util.List;

/**
 * 标签 ViewModel
 */
public class TagViewModel extends ViewModel {
    
    private final TagRepository tagRepository;
    private final LiveData<List<Tag>> tags;
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    
    public TagViewModel(TagRepository tagRepository) {
        this.tagRepository = tagRepository;
        this.tags = tagRepository.getAllTags();
    }
    
    public LiveData<List<Tag>> getTags() {
        return tags;
    }
    
    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }
    
    public void insertTag(Tag tag) {
        isLoading.setValue(true);
        tagRepository.insertTag(tag);
        isLoading.setValue(false);
    }
    
    public void updateTag(Tag tag) {
        isLoading.setValue(true);
        tagRepository.updateTag(tag);
        isLoading.setValue(false);
    }
    
    public void deleteTag(long tagId) {
        isLoading.setValue(true);
        tagRepository.deleteTag(tagId);
        isLoading.setValue(false);
    }
    
    public static class Factory implements ViewModelProvider.Factory {
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(Class<T> modelClass) {
            if (modelClass.isAssignableFrom(TagViewModel.class)) {
                NoteAIApplication application = NoteAIApplication.getInstance();
                return (T) new TagViewModel(application.getTagRepository());
            }
            throw new IllegalArgumentException("Unknown ViewModel class");
        }
    }
}

File: app/src/main/java/com/notai/ui/viewmodel/TagViewModel.kt
package com.notai.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.notai.NoteAIApplication
import com.notai.data.model.Tag
import com.notai.data.repository.TagRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * 标签 ViewModel
 */
class TagViewModel(
    private val tagRepository: TagRepository
) : ViewModel() {
    
    private val _tags = MutableStateFlow<List<Tag>>(emptyList())
    val tags: StateFlow<List<Tag>> = _tags.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    init {
        loadTags()
    }
    
    fun loadTags() {
        viewModelScope.launch {
            tagRepository.getAllTags().collect { tagsList ->
                _tags.value = tagsList
            }
        }
    }
    
    fun insertTag(tag: Tag) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                tagRepository.insertTag(tag)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun updateTag(tag: Tag) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                tagRepository.updateTag(tag)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun deleteTag(tagId: Long) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                tagRepository.deleteTag(tagId)
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    companion object {
        fun provideFactory(): ViewModelProvider.Factory {
            return object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T {
                    val application = NoteAIApplication.getInstance()
                    return TagViewModel(application.tagRepository) as T
                }
            }
        }
    }
}

